caddyhttp/fastcgi/fastcgi_test.go

```diff
@@ -51,6 +52,106 @@ func TestServeHTTP(t *testing.T) {
 	}
 }
 
+// listenerWithConnectionCounter adds a counter to keep track
+// of the number of accepted connections.
+type listenerWithConnectionCounter struct {
+	net.Listener
+	sync.RWMutex
+	ConnectionsAcceptedCounter int
+}
+
+func (l *listenerWithConnectionCounter) Accept() (net.Conn, error) {
+	l.Lock()
+	l.ConnectionsAcceptedCounter++
+	l.Unlock()
+	return l.Listener.Accept()
+}
+
+// TestPersistent ensures that persistent
+// as well as the non-persistent fastCGI servers
+// send the answers corresnponding to the correct request.
+// It also checks the number of tcp connections used.
+func TestPersistent(t *testing.T) {
+	numberOfRequests := 32
+
+	for _, poolsize := range []int{0, 1, 5, numberOfRequests} {
+		l, err := net.Listen("tcp", "127.0.0.1:0")
+		if err != nil {
+			t.Fatalf("Unable to create listener for test: %v", err)
+		}
+
+		listener := &listenerWithConnectionCounter{l, *new(sync.RWMutex), 0}
+
+		// this fcgi server replies with the request URL
+		go fcgi.Serve(listener, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			body := "This answers a request to " + r.URL.Path
+			bodyLenStr := strconv.Itoa(len(body))
+
+			w.Header().Set("Content-Length", bodyLenStr)
+			w.Write([]byte(body))
+		}))
+
+		network, address := parseAddress(listener.Addr().String())
+		handler := Handler{
+			Next:  nil,
+			Rules: []Rule{{Path: "/", Address: listener.Addr().String(), dialer: &persistentDialer{size: poolsize, network: network, address: address}}},
+		}
+
+		sem := make(chan int, numberOfRequests)
+		serialMutex := new(sync.Mutex)
+
+		serialCounter := 0
+		parallelCounter := 0
+		// make some serial followed by some
+		// parallel requests to challenge the handler
+		for _, serialize := range []bool{true, false, false, false} {
+			if serialize {
+				serialCounter++
+			} else {
+				parallelCounter++
+			}
+
+			for i := 0; i < numberOfRequests; i++ {
+				go func(i int) {
+					if serialize {
+						serialMutex.Lock()
+						defer serialMutex.Unlock()
+					}
+					r, err := http.NewRequest("GET", "/"+strconv.Itoa(i), nil)
+					if err != nil {
+						t.Errorf("Unable to create request: %v", err)
+					}
+					w := httptest.NewRecorder()
+
+					status, err := handler.ServeHTTP(w, r)
+
+					if status != 0 {
+						t.Errorf("Handler(pool: %v) return status %v", poolsize, status)
+					}
+					if err != nil {
+						t.Errorf("Handler(pool: %v) Error: %v", poolsize, err)
+					}
+					want := "This answers a request to /" + strconv.Itoa(i)
+					if got := w.Body.String(); got != want {
+						t.Errorf("Expected response from handler(pool: %v) to be '%s', got: '%s'", poolsize, want, got)
+					}
+					sem <- 0 // signal semaphore
+				}(i)
+			} //next request
+			// wait for semaphore
+			for j := 0; j < numberOfRequests; j++ {
+				<-sem
+			}
+		} // next set of requests (serial/parallel)
+
+		listener.Close()
+		listener.RLock()
```

Comment:
That's right, there is currently no need for a lock since all other threads have finished in this test. I thought it might be good hygiene to lock this field anyway, for any access, in view of possible future changes.